### PR TITLE
Type argument inference fix for infinitely recursive anonymous types

### LINKED ISSUE #3451: Stack overflow with type argument inference on infinitely recursive anonymous type
This is very similar to #3309. It just uses type argument inference instead of assignability.
```ts
function foo<T>() {
    var z = foo<typeof y>();
    var y: {
        y2: typeof z
    };
    return y;
}


function bar<T>() {
    var z = bar<typeof y>();
    var y: {
        y2: typeof z;
    }
    return y;
}

var a = foo<number>();
var b = bar<number>();

function test<T>(x: typeof a): void { }
test(b);
```

### PR DESCRIPTION
**Pull request by JsonFreeman**
Type argument inference should use the same infrastructure as assignability when it comes to detecting infinite recursion in a type structure. This change makes type argument inference rely on the same isDeeplyNestedGeneric function.

Fixes #3451.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -5210,23 +5195,27 @@ module ts {
...
+                    targetStack[depth] = target;
+                    depth++;
+                    inferFromProperties(source, target);
+                    inferFromSignatures(source, target, SignatureKind.Call);
+                    inferFromSignatures(source, target, SignatureKind.Construct);
+                    inferFromIndexTypes(source, target, IndexKind.String, IndexKind.String);
+                    inferFromIndexTypes(source, target, IndexKind.Number, IndexKind.Number);
+                    inferFromIndexTypes(source, target, IndexKind.String, IndexKind.Number);
+                    depth--;                }
             }
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: What is with this brace?

Comment (JsonFreeman) [inline src/compiler/checker.ts]: That is a white space management error

**Comment by JsonFreeman** [comment]
@ahejlsberg Do you have any feedback on this?

**Comment by ahejlsberg** [comment]
Overall, I like this unification. We used to stop after five levels of nesting during type inference, but now we'll go to ten levels. I doubt there are real world examples where this makes a difference in the quality of the inferences, but it certainly might impact performance (e.g. we'll spend time making the same inference ten times instead of five times). Thoughts? Perhaps we should pass the limit as a parameter so we can keep it at five for inference?

**Comment by JsonFreeman** [comment]
Passing a limit parameter is an option. And I agree that the quality of the inference will not be impacted in realistic cases.

That said, if we are concerned about performance, and we think changing the 10 to a 5 will help, I actually think we should do it across the board, for assignability too. Assignability is way more common than type argument inference, and in fact, I think every round of type argument inference is followed by a round of assignability checking. So we'd get way more mileage if we did that change across the board, and we'd get to keep the limit consistent. I also don't think we would sacrifice much correctness if we change to 5 for assignability.

**Comment by ahejlsberg** [comment]
Yeah, changing to 5 across the board is probably better. Let's do that.

**Pull request merged**